accepted/2021/json-polymorphism.md

```diff
@@ -0,0 +1,275 @@
+}
+```
+Currently, when serializing a `Bar` instance as type `Foo`
+the serializer will apply the JSON schema derived from the type `Foo`:
+```csharp
+Foo foo1 = new Foo { A = 1 };
+Foo foo2 = new Bar { A = 1, B = 2 };
+Foo foo3 = new Baz { A = 1, B = 2, C = 3 };
+
+JsonSerializer.Serialize<Foo>(foo1); // { "A" : 1 }
+JsonSerializer.Serialize<Foo>(foo2); // { "A" : 1 }
+JsonSerializer.Serialize<Foo>(foo3); // { "A" : 1 }
+```
+Under the new proposal we can change this behaviour by annotating 
+the base class (or interface) with the `JsonPolymorphicType` attribute:
+```csharp
+[JsonPolymorhicType]
+public class Foo
+{
+  ...
+}
+```
+which will result in the above values now being serialized as follows:
+```csharp
+JsonSerializer.Serialize<Foo>(foo1); // { "A" : 1 }
+JsonSerializer.Serialize<Foo>(foo2); // { "A" : 1, "B" : 2 }
+JsonSerializer.Serialize<Foo>(foo3); // { "A" : 1, "B" : 2, "C" : 3 }
+```
+Note that the `JsonPolymorphic` attribute is not inherited by derived types.
+In the above example `Bar` inherits from `Foo` yet is not polymorphic in its own right:
+```csharp
+Bar bar = new Baz { A = 1, B = 2, C = 3 };
+JsonSerializer.Serialize<Bar>(bar); // { "A" : 1, "B" : 2 }
+```
+If annotating the base class with an attribute is not possible,
+polymorphism can alternatively be opted in for a type using the
+new `JsonSerializerOptions.SupportedPolymorphicTypes` predicate:
+```csharp
+public class JsonSerializerOptions
+{
+  public Func<Type, bool> SupportedPolymorphicTypes { get; set; }
+}
+```
+Applied to the example above:
+```csharp
+var options = new JsonSerializerOptions { SupportedPolymorphicTypes = type => type == typeof(Foo) };
+JsonSerializer.Serialize<Foo>(foo1, options); // { "A" : 1, "B" : 2 }
+JsonSerializer.Serialize<Foo>(foo2, options); // { "A" : 1, "B" : 2, "C" : 3 }
+```
+It is always possible to use this setting to enable polymorphism _for every_ serialized type:
+```csharp
+var options = new JsonSerializerOptions { SupportedPolymorphicTypes = _ => true };
+
+// `options` treats both `Foo` and `Bar` members as polymorphic
+Baz baz = new Baz { A = 1, B = 2, C = 3 };
+JsonSerializer.Serialize<Foo>(baz, options); // { "A" : 1, "B" : 2, "C" : 3 }
+JsonSerializer.Serialize<Bar>(baz, options); // { "A" : 1, "B" : 2, "C" : 3 }
+```
+As mentioned previously, this feature provides no provision for deserialization.
+If deserialization is a requirement, users would need to opt for the
+polymorphic serialization with type discriminators feature.
+
+## Polymorphism with type discriminators
+
+This feature allows users to opt in to polymorphic serialization for a given type
+by associating string identifiers with particular subtypes in the hierarchy.
+These identifiers are written to the wire so this brand of polymorphism is roundtrippable.
+
+At the core of the design is the introduction of `JsonKnownType` attribute that can
+be applied to type hierarchies like so:
+```csharp
+[JsonKnownType(typeof(Derived1), "derived1")]
```

Comment:
Would users be able to specify the name of the discriminator field in the serialized model here? The examples use `$type`, but not all web APIs which use discriminated union-like types use that naming.

Reply:
> Would users be able to specify the name of the discriminator field in the serialized model here? The examples use $type, but not all web APIs which use discriminated union-like types use that naming.

Yes, that is something we are considering although it has not been included in the prototype for now. There are a few infrastructural problems to be sorted out, for example our metadata parsing implementation currently requires that all metadata properties start with `$` and are placed before any other properties in the object.

> Furthermore: are non-string values permitted? It's also often the case that web APIs use integers for these discriminator values, which would often be mapped to an enum in C# code.

Not in the prototype, but you're making a good point that this should be supported. We would probably need to make the `JsonKnownTypeAttribute` accept `object` parameters but require that each type uses a unique identifier type, so the following configuration would be invalid:
```csharp
[JsonKnownType(typeof(Bar), 1)]
[JsonKnownType(typeof(Baz), "baz")]
public class Foo {}
```
As far as supported identifier types go, am I right to assume that only `string` and `int` values should be accepted?

Reply:
> for example our metadata parsing implementation currently requires that all metadata properties start with `$` and are placed before any other properties in the object.

Yeah - in my own converters I've resorted to making a copy of the reader, finding and deserializing the discriminator, and then using the reader copy to deserialize as the actual type, updating the original reader with the copy after the fact. While this fixes the "must be the first field in the JSON object" issue, it's not that great because it parses the value multiple times.

> As far as supported identifier types go, am I right to assume that only string and int values should be accepted?

As far as I'm aware, yes, only `string` and `int` are used for these. However, I think it may be convenient to support enum types too, as they're easily converted to `int`, and are (IMO) the natural type to use for a discriminator.